File: day4/solution.py
with open('input.txt', 'r') as f:
    lines = f.readlines()

lines = [x.strip() for x in lines]
# Question 1: 
solution = 0
for pair in lines:
    elf1, elf2 = pair.split(',')
    elf1_start, elf1_end = elf1.split('-')
    elf2_start, elf2_end = elf2.split('-')

    if ((int(elf1_start) <= int(elf2_start)) and (int(elf1_end) >= int(elf2_end))) or ((int(elf1_start) >= int(elf2_start)) and (int(elf1_end) <= int(elf2_end))):
        solution += 1

print('Question 1: ', solution)

# Question 2:
solution = 0 
for pair in lines:
    elf1, elf2 = pair.split(',')
    elf1_start, elf1_end = elf1.split('-')
    elf2_start, elf2_end = elf2.split('-')

    if (int(elf2_start) <= int(elf1_start) <= int(elf2_end)) or (int(elf1_start) <= int(elf2_start) <= int(elf1_end)):
        solution += 1

print('Question 2: ', solution)

File: day4/solution.js
const fs = require("fs");

let data = fs.readFileSync("input.txt", "utf8");
data = data.split("\n").filter((item) => item);

// Question 1:

const solution1 = data.reduce((agg, curr) => {
  const [elf1, elf2] = curr.split(",");
  const [elf1Start, elf1End] = elf1.split("-").map((x) => Number(x));
  const [elf2Start, elf2End] = elf2.split("-").map((x) => Number(x));
  if (elf1Start <= elf2Start && elf1End >= elf2End) return agg + 1;
  if (elf2Start <= elf1Start && elf2End >= elf1End) return agg + 1;
  return agg;
}, 0);

console.log("Question 1: ", solution1);

// Question 2:

const solution2 = data.reduce((agg, curr) => {
  const [elf1, elf2] = curr.split(",");
  const [elf1Start, elf1End] = elf1.split("-").map((x) => Number(x));
  const [elf2Start, elf2End] = elf2.split("-").map((x) => Number(x));
  if (elf1Start <= elf2Start && elf1End >= elf2Start) return agg + 1;
  if (elf2Start <= elf1Start && elf2End >= elf1Start) return agg + 1;
  return agg;
}, 0);

console.log("Question 2: ", solution2);
